shiny/ui/_sidebar.py

```diff
@@ -52,94 +117,270 @@ class directly. Instead, supply the :func:`~shiny.ui.sidebar` object to
+        interpreted as pixels) or a character vector with valid CSS lengths. `padding`
+        may be one to four values.
+
+        * If a single value, then that value will be used for all four sides.
+        * If two, then the first value will be used for the top and bottom, while
+          the second value will be used for left and right.
+        * If three values, then the first will be used for top, the second will be left
+          and right, and the third will be bottom.
+        * If four, then the values will be interpreted as top, right, bottom, and left
+          respectively.
     """
 
     def __init__(
         self,
-        tag: Tag,
-        collapse_tag: Optional[Tag],
-        position: Literal["left", "right"],
-        open: Literal["desktop", "open", "closed", "always"],
-        width: CssUnit,
-        max_height_mobile: Optional[str | float],
-        color_fg: Optional[str],
-        color_bg: Optional[str],
+        children: tuple[TagChild | TagAttrs, ...],
+        attributes: dict[str, TagAttrValue],
+        position: Literal["left", "right"] = "left",
+        open: Optional[SidebarOpenValues | SidebarOpen] = None,
+        width: CssUnit = 250,
+        id: Optional[str] = None,
+        title: TagChild | str = None,
+        fg: Optional[str] = None,
+        bg: Optional[str] = None,
+        class_: Optional[str] = None,
+        max_height_mobile: Optional[str | float] = None,
+        gap: Optional[CssUnit] = None,
+        padding: Optional[CssUnit | list[CssUnit]] = None,
     ):
-        self.tag = tag
-        self.collapse_tag = collapse_tag
+        if isinstance(title, (str, int, float)):
+            title = tags.header(str(title), class_="sidebar-title")
+
+        self.id: ResolvedId | None = resolve_id_or_none(id)
+        self.title = title
+        self.class_ = class_
+        self.gap = as_css_unit(gap)
+        self.padding = as_css_padding(padding)
+        self._open: SidebarOpen | None = self._as_open(open)
+        self._default_open = SidebarOpen()
         self.position = position
-        self.open = open
-        self.width = width
-        self.max_height_mobile = max_height_mobile
-        self.color_fg = color_fg
-        self.color_bg = color_bg
-
-    # The `Sidebar` class should use it's fields, not this method
-    def tagify(self) -> Tag:
-        """
-        Not implemented
-        """
-        # Similar to `NavMenu.tagify()`
-        raise NotImplementedError(
-            "`Sidebar` objects must be handled by `layout_sidebar(*args)`."
+        self.width = as_css_unit(width)
+        self._max_height_mobile = max_height_mobile
+        self.color = {"fg": fg, "bg": bg}
+        self.attributes = attributes
+        self.children = children
+
+    @property
+    def open(self) -> SidebarOpen:
+        if self._open is None:
+            return self._default_open
+
+        return self._open
```

Comment:
I'm generally suspicious of using dynamic properties because I've been burned by surprising behavior from them a number of times, and I think in general they make it harder to reason about the code.

Instead of (conditionally) setting `self._open` here, can that be done in the constructor?

Reply:
Unfortunately, we can't do it in the constructor because we need the `sidebar` object to carry the signal that the user hasn't supplied an explicit decision about `open`.

I set this up as a property so that consumers of `sidebar` can just query `sidebar.open` to get the resolved value, or can consult the private `sidebar._open` property. Otherwise consumers have to manually fill in the details.

Reply:
If the goal is to signal whether the user is using the default, it would be simpler and easier to understand if there was a property or method to indicate that -- something like `Sidebar.open_is_default`.

As it is in the code here, one would have to know that when `._open` is `None`, that means the user is just using the default, but that fact is not obvious. (Also, I don't see that we use the default/not default information anywhere, but perhaps it's for some future use?)

With respect to the dynamic setter, it's surprising that the value can be automatically changed on assignment:

```py
s = ui.sidebar()
s.open = "open"
s.open
#> SidebarOpen(desktop='open', mobile='open', _VALUES=('open', 'closed', 'always'))
```

It's less confusing to to just have the developer assign the full object, if they want to change `s.open`. That seems like a rare situation, but if it were more common, we could provide a setter which does automatic upgrading, like `s.set_open("open")`.

So I think we should:
- Make `.open` a normal property instead of a dynamic one.
- Set `.open` in the constructor (and use a default value if the `open` argument is `None`)
- If needed, add a `.open_is_default` property (or method).
- If needed, add a `.set_open()` function which can accept strings and "upgrade" them.

Reply:
Another quick thought:

- If another component wants to set value `open` only if the user hasn't set it, we could add a function like `Sidebar.set_open_if_not_set("open")`  or maybe `Sidebar.ensure_open("open")`.

Reply:
> * If another component wants to set value `open` only if the user hasn't set it, we could add a function like `Sidebar.set_open_if_not_set("open")`  or maybe `Sidebar.ensure_open("open")`.

Thanks @wch -- we already have `Sidebar._set_default_open()` which does exactly this with the additional benefit of returning a new `Sidebar` object rather than mutating in place.

My answer on Friday was probably too short because I was answering on my phone. Consulting the private `._open` method is an implementation detail and not part of the public API. The public API is handled through the `.open` property. As a `@property`, any consumers of the `Sidebar` instance can just look at `.open` to get an expected and consistent value. Users can update its value with `<sidebar>.open = <new_value>`. Internally, if we want to change behavior, higher-level functions call `Sidebar._set_default_open(<new_value>)` without having to worry about the value of `._open`.

Overall, I think there may have been a misunderstanding about the relative importance of the "signal" that the user hasn't supplied the default value. I only meant that we internally needed to track that, not that we need to make it clear outside of the `Sidebar` class. To me, it's far more important that the rest of our code can call `<sidebar>.open.desktop` knowing that `open` will return a dictionary with `desktop` and `mobile`.

I can understand wanting to force anyone who updates `<sidebar>.open` post hoc to provide a fully realized object, but to me it's also very relevant that `ui.Sidebar` is really the implementation of `ui.sidebar()`. Previously, some of the implementation happened in the `sidebar()` function and other parts happened in the `Sidebar` class. This PR moves _everything_ other than collecting children and attrs into the `Sidebar` class. We could let `ui.sidebar()` take scalar strings for `open`, but throw if `Sidebar.open` receives a scalar string, but this feels overly strict to me. I think it's worth it to keep `ui.sidebar()` extremely thin around `ui.Sidebar`.

Reply:
After talking through this with @wch, we decided to make `Sidebar.open()` a method rather than using a `@property`. This makes it clearer from the outside that some work is being done to provide the `open` value. The method can also set the sidebar's `open` value.

```python
from shiny import ui
sidebar = ui.sidebar()
sidebar.open()
#> SidebarOpen(desktop='open', mobile='closed')

sidebar.open("closed")
#> SidebarOpen(desktop='closed', mobile='closed')

sidebar.open({"mobile": "always"})
#> SidebarOpen(desktop='open', mobile='always')

sidebar.open(None) # reset to use default
#> SidebarOpen(desktop='open', mobile='closed')

sidebar2 = sidebar._set_default_open("always")
sidebar.open()
#> SidebarOpen(desktop='open', mobile='closed')
sidebar2.open()
#> SidebarOpen(desktop='always', mobile='always')
```